### PR TITLE
implement common http transport for modules with http method

### LINKED ISSUE #37: sockets leak when fetching module with verify=false
### Description
`exporter_exporter` leaks sockets when serving metrics for module with `verify=false`.

### How to reproduce
Run `exporter_exporter` with some module configured like this:
```yaml
modules:
  nodeNoVerify:
    method: http
    http:
      verify: false
      port: 9100
```
Inspect current statistics on sockets in your system:
```bash
$ ss -s
Total: 1295
TCP:   71 (estab 20, closed 16, orphaned 0, timewait 2)

Transport Total     IP        IPv6
RAW	  1         0         1        
UDP	  31        18        13       
TCP	  55        38        17       
INET	  87        56        31       
FRAG	  0         0         0        
```
Generate some load and check statistics again e.g.:
```bash
$ go-wrk -c=4 -n=2000 http://localhost:9999/proxy?module=nodeNoVerify
...
$ ss -s
Total: 3329
TCP:   6283 (estab 2054, closed 4193, orphaned 0, timewait 4180)

Transport Total     IP        IPv6
RAW	  1         0         1        
UDP	  31        18        13       
TCP	  2090      1054      1036     
INET	  2122      1072      1050     
FRAG	  0         0         0  
```
Checkout lots of open connections between `exporter_exporter` and `node_exporter`:
```bash
$ ss -etnp | grep 9100
```
Checkout logs of `exporter_exporter`:
```
2020/01/15 12:15:01 http: Accept error: accept tcp [::]:9999: accept4: too many open files; retrying in 5ms
2020/01/15 12:15:01 http: proxy error: dial tcp: lookup localhost: device or resource busy
```

### Expected behavior
One open connection to `node_exporter`

### Actual behavior
Approx. 2000 open connections to `node_exporter`

### Possible solution
This happens because `http.Transport` struct must be reused (as doc suggests) but new instance is created every time `httpConfig.ServerHTTP` method is invoked. Creating `http.Transport` with `DisableKeepAlives: true` also solves the issue

### PR DESCRIPTION
**Pull request by cyril-s**
Current implementation handles modules with `verify=true` and `verify=false` very differently. For the former, it uses `promhttp.HandlerFor` helper with `http.Client` and acts as some kind of gateway while for the latter, it uses `httputil.ReverseProxy` and acts as usual http proxy. Some problems are:

1) Inconsistent http headers handling. `http.Client` makes request anew while `httputil.ReverseProxy` copies end-to-end headers
2) Inconsistent error handling. `promhttp.HandlerFor` always returns `500 Internal Server Error` when `302` and `304` codes are appropriate, while `promhttp.HandlerFor` of current implementation returns `302` always.

I suggest get rid of `promhttp.HandlerFor` and move verification logic into `ModifyResponse` hook  of `httputil.ReverseProxy`. This allows for usual http proxy headers handling and better error handling.

Error handling changes table:

| | Old verify | Old no verify | New verify | New no verify |
|-|------------|---------------|------------|---------------|
| Bad format | 500, empty body | 200 | 500, error message in body |200 |
| Timeout | 500, “context deadline exceeded” in body | 502, empty body | 504, Gateway Timeout | 504, Gateway Timeout |
| Network error | 500, “connection refused” | 502, empty body | 502, Bad Gateway | 502, Bad Gateway |
| Default | 500, some detailed error message | 502, empty body | 502, Bad Gateway | 502, Bad Gateway |

In summary, this PR changes:
* http modules now embed httputil.ReverseProxy
* GatherWithContext and usage of promhttp.HandlerFor are dropped
  and verification logic is moved into ModifyResponse hook of
  httputil.ReverseProxy
* Consistent error handling is implemented in ErrorHandler hook
  of httputil.ReverseProxy
* added benchmark for reverse proxy handler

Closes #37

### CODE REVIEW

**Comment by tcolgate** [comment]
This looks like an excellent approach. I'll need a little time to review to PR though.

**Comment by tcolgate** [comment]
What happen in the case of a 401 (or other non-200/500)?

File: http.go [hunk 1]
```diff
@@ -14,110 +14,115 @@
...
+
+		r.URL.Scheme = cfg.HTTP.Scheme
+		r.URL.Host = net.JoinHostPort(cfg.HTTP.Address, strconv.Itoa(cfg.HTTP.Port))
+		r.URL.Path = cfg.HTTP.Path
+	}
+}
 
+func (cfg moduleConfig) getReverseProxyModifyResponseFunc() func(*http.Response) error {
+	return func(resp *http.Response) error {
 		if resp.StatusCode != 200 {
```
Comment (tcolgate) [inline http.go]: I may be missing it, but can we not log non-200 responses here? Are the logged elsewhere? Without it, that could make debugging things a bit tricky.

**Comment by tcolgate** [comment]
Generally happy, but I'd like to keep reporting of non-200 responses if possible. Debugging things like 401s, and/or https issues needs to be as easy as possible. I've not had time to read the code, so I apologise if this is dealt with and I'm missing it. 
  Thanks especially for the benchmark tests.

**Comment by cyril-s** [comment]
@tcolgate only errors originating from exporter_exporter itself are logged. They are:
* failure to decode metrics when `Verify: true`
* Network errors e.g. timeout, connection refused etc.

All other errors are passed as is from proxied server to the client. But I agree that logging would be useful. I made another PR with access logging. Take a look at #40

**Comment by tcolgate** [comment]
okay, we can review logging in the other PR.
